### DISCUSSION TITLE
Change LocalizedString at runtime

### DISCUSSION THREAD

**Discussion by AndyCaignie** [xamarin/XamarinCommunityToolkit#1627] [Q&A] [answered]
Hello,
I'm trying to make my app multi language using the LocalizedString.
However I have one major issue: some texts change at runtime. In the tutorial the Version {0} is used, but the formatted string is a constant. I need this to be filled in at runtime, as it displays a value that is retrieved after creation of the LocalizedString.
How can I change the text of the LocalizedString after it is created?

Thx in advance,
Andy

**Comment by maxkoshevoi** [comment, marked as answer]
Hi, this video and sample repo might be of some assistance:

- https://www.youtube.com/watch?v=wyKKa09z9VQ
- https://github.com/maxkoshevoi/XCTLocalizationDemo

**Reply by AndyCaignie** [reply]
Hi Maksym, I have watched the (great!) video to start with my localisation.
But my issue is that the text of a label, button, .. changes at runtime. Suppose I would have a label displaying the open messages, I could use something like

public LocalizedString OpenMessage { get; } = new(() => string.Format(Translations.OpenMessages, nbrOfOpenMessages));

But how do I change the OpenMessage if nbrOfOpenMessages changes while the page is open?
Or in your repo: what if AppInfo.VersionString would change at runtime? How is the AppVersion string then refreshed?

**Reply by maxkoshevoi** [reply]
> my issue is that the text of a label, button, .. changes at runtime

This is not an issue at all! `LocalizedString` was designed exactly for this purpose. Notice, that when you create `LocalizedString`, you supply a function to it:

```cs
public LocalizedString OpenMessage { get; } = new(() => string.Format(Translations.OpenMessages, nbrOfOpenMessages));
```

When you change the language, it reruns that function and updates it's value. The only thing that is required from your side is to change the language using `LocalizationResourceManager` (this will trigger an even that all `LocalizedString`s are subscribed to).

Also note that changing the language through `LocalizationResourceManager` doesn't automatically change it for your resources (`Translations`), so you also need to write something like this: https://github.com/maxkoshevoi/XCTLocalizationDemo/blob/main/XCTLocalizationDemo/XCTLocalizationDemo/App.xaml.cs#L13

```cs
LocalizationResourceManager.Current.PropertyChanged += (_, _) => Translations.Culture = LocalizationResourceManager.Current.CurrentCulture;
```

> Or in your repo: what if AppInfo.VersionString would change at runtime?

In the provided demo project displayed value of `AppInfo.VersionString` would be automatically changed at runtime via mechanisms described above.

**Reply by AndyCaignie** [reply]
Hello Maksym, thx for the clarification but I cannot get this working :-(. If I change the language then the function is indeed rerun, but not if nbrOfOpenMessages changes. I changed the string.format to a function but that didn't change the behaviour.

**Reply by maxkoshevoi** [reply]
Could you share a repro project?

**Reply by maxkoshevoi** [reply]
Issue in your case that language isn't changed. You change parts of the resulting string, and `LocalizedString` doesn't know about it, so it cannot re-render itself. You need to call `OnPropertyChanged(nameof(CounterText))` after `Counter` is updated to notify view that `LocalizedString` needs to be re-rendered.
I've created two PRs in your repo to demonstrate possible fixes.
(https://github.com/AndyCaignie/TranslationProblem/pull/1 https://github.com/AndyCaignie/TranslationProblem/pull/2)

**Reply by AndyCaignie** [reply]
Hello Maksym thank you very much for your help, this works perfectly!

**Comment by AndyCaignie** [comment]
I created a small XAMARIN project on git: https://github.com/AndyCaignie/TranslationProblem
The about page is changed with 2 labels, when I push the button to add to the counter I would like both labels to update instead of just the second one
